runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -114,7 +144,525 @@ native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
+#if !defined(J9VM_ENV_LITTLE_ENDIAN)
+				/* Right shift the 64-bit float argument by 4 bytes(32 bits) given the actual value
+				 * is placed on the higher 4 bytes on the Big-Endian(BE) platforms.
+				 */
+				if (J9_FFI_UPCALL_SIG_TYPE_FLOAT == argSigType) {
+					argValue = argValue >> J9_FFI_UPCALL_SIG_TYPE_32_BIT;
+				}
+#endif /* J9VM_ENV_LITTLE_ENDIAN */
+				*(I_32*)--currentThread->sp = (I_32)argValue;
+				break;
+			}
+			case J9_FFI_UPCALL_SIG_TYPE_INT64: /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+				currentThread->sp -= 2;
+				*(I_64*)currentThread->sp = *(I_64*)getArgPointer(nativeSig, argsListPointer, argIndex);
+				break;
+			case J9_FFI_UPCALL_SIG_TYPE_POINTER:
+			{
+				I_64 offset = *(I_64*)getArgPointer(nativeSig, argsListPointer, argIndex);
+				j9object_t memAddrObject = createMemAddressObject(data, offset, isCurThrdAllocated);
+				if (NULL == memAddrObject) {
+					/* The OOM exception set in createMemAddressObject() will be thrown in the interpreter
+					 * after returning from the native function in downcall.
+					 */
+					throwOOM = true;
+					goto done;
+				}
+				 /* Push the object on the special frame so as to avoid updating the address on the stack
+				  * by GC when allocating memory for the next pointer/struct of the argument list.
+				  */
+				PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, memAddrObject);
+				break;
+			}
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT:
+			{
+				I_64 offset = (I_64)(intptr_t)getArgPointer(nativeSig, argsListPointer, argIndex);
+				j9object_t memSegmtObject = createMemSegmentObject(data, offset, sigArray[argIndex].sizeInByte, isCurThrdAllocated, sessionOrScopeObject);
+				if (NULL == memSegmtObject) {
+					/* The OOM exception set in createMemSegmentObject() will be thrown in the interpreter
+					 * after returning from the native function in downcall.
+					 */
+					throwOOM = true;
+					goto done;
+				}
+				 /* Push the object on the special frame so as to avoid updating the address on the stack
+				  * by GC when allocating memory for the next pointer/struct of the argument list.
+				  */
+				PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, memSegmtObject);
+				break;
+			}
+			default:
+				Assert_VM_unreachable();
+				break;
+			}
+		}
+		/* Only restore the literals and keep the unchanged arguments on the stack which
+		 * are passed over to native2InterpreterTransition() in the interpreter so as to
+		 * invoke the upcall method handle.
+		 */
+		currentThread->literals = thrLiterals;
+
+		/* Place mhMetaData as the return value to native2InterpreterTransition() when calling into
+		 * the interpreter so as to set the invoke cache array (MemberName and appendix)
+		 * before invoking the target handle in upcall.
+		 */
+		currentThread->returnValue = J9_BCLOOP_N2I_TRANSITION;
+		currentThread->returnValue2 = (UDATA)data;
+		c_cInterpreter(currentThread);
+
+done:
+		restoreCallInFrameHelper(currentThread);
```

Comment:
do we need to do an exception check before returning to native?

Reply:
The exception check has been done at `inlInternalDowncallHandlerInvokeNative` when returning back to the interpreter:
```
/runtime/vm/BytecodeInterpreter.hpp
	inlInternalDowncallHandlerInvokeNative(REGISTER_ARGS_LIST)
	{
...
		ffi_call(cif, FFI_FN(function), returnStorage, values);
#endif /* FFI_NATIVE_RAW_API */
		VM_VMHelpers::afterJNICall(_currentThread);
		VM_VMAccess::inlineEnterVMFromJNI(_currentThread);
		VMStructHasBeenUpdated(REGISTER_ARGS);
#if JAVA_SPEC_VERSION >= 19
		_currentThread->callOutCount -= 1;
#endif /* JAVA_SPEC_VERSION >= 19 */
		if (VM_VMHelpers::exceptionPending(_currentThread)) { <------
			rc = GOTO_THROW_CURRENT_EXCEPTION;
		}
```
In addition, there is no way to check the exception if the native thread is created locally (which is not the same thread in downcall), which means the thread will be cleaned up once the dispatcher exits and returns to the interpreter.

Reply:
If the downcall thread can be stored in upcall meta data as mentioned at https://github.com/eclipse-openj9/openj9/pull/15309#discussion_r960088631, it might be possible to transfer the exception from the native thread (created locally) to the downcall thread before returning to the interpreter.